base/iterators.jl

```diff
@@ -924,47 +926,97 @@ julia> collect(Iterators.partition([1,2,3,4,5], 2))
  [1, 2]
  [3, 4]
  [5]
+
+julia> collect(Iterators.partition([1,2,3,4,5], 2; flush=true))
+2-element Array{Array{Int64,1},1}:
+ [1, 2]
+ [3, 4]
+
+julia> collect(Iterators.partition([1,2,3,4,5], 2; step=1))
+2-element Array{Array{Int64,1},1}:
+ [1, 2]
+ [2, 3]
+ [3, 4]
+ [4, 5]
 ```
 """
-partition(c::T, n::Integer) where {T} = PartitionIterator{T}(c, Int(n))
-
+function partition(c::T, n::Integer;
+                   step::Integer = n, flush::Bool = true) where {T}
+    n < 1 && throw(ArgumentError("n (= $n) must be larger than 1"))
+    step < 1 && throw(ArgumentError("step (= $step) must be larger than 1"))
+    if T <: AbstractVector
+        return PartitionIterator{T,Nothing}(c, Int(n), Int(step), flush)
+    elseif step >= n
+        return PartitionIterator{T,:GE}(c, Int(n), Int(step), flush)
+    else
+        return PartitionIterator{T,:LESS}(c, Int(n), Int(step), flush)
+    end
+end
 
-struct PartitionIterator{T}
+struct PartitionIterator{T,STEP}
     c::T
     n::Int
+    step::Int
+    flush::Bool
 end
 
-eltype(::Type{PartitionIterator{T}}) where {T} = Vector{eltype(T)}
+eltype(::Type{<:PartitionIterator{T}}) where {T} = Vector{eltype(T)}
 partition_iteratorsize(::HasShape) = HasLength()
 partition_iteratorsize(isz) = isz
-function IteratorSize(::Type{PartitionIterator{T}}) where {T}
+function IteratorSize(::Type{<:PartitionIterator{T}}) where {T}
     partition_iteratorsize(IteratorSize(T))
 end
 
 IteratorEltype(::Type{<:PartitionIterator{T}}) where {T} = IteratorEltype(T)
 
 function length(itr::PartitionIterator)
     l = length(itr.c)
-    return div(l, itr.n) + ((mod(l, itr.n) > 0) ? 1 : 0)
+    if l <= max(itr.step, itr.n)
+        return l > 0 && (itr.n <= l || itr.flush) ? 1 : 0
+    elseif itr.step >= itr.n
+        return div(l, itr.step) + (itr.flush && (mod(l, itr.step) > 0) ? 1 : 0)
+    else
+        m = l - itr.n
+        return div(m, itr.step) + (itr.flush && (mod(m, itr.step) > 0) ? 1 : 0) + 1
+    end
 end
 
-function iterate(itr::PartitionIterator{<:Vector}, state=1)
+function iterate(itr::PartitionIterator{<:AbstractVector}, state=1)
```

Comment:
Note the change in the type signature. I suggest broadening the types to be matched with this implementation as it's pretty easy to use the iterator-based fall-back implementation with `Base.Generator(identity, myarray)`.